rpcs3/Crypto/aes.cpp

```diff
@@ -558,6 +562,13 @@ int aes_setkey_dec( aes_context *ctx, const unsigned char *key, unsigned int key
     if( ret != 0 )
         return( ret );
 
+    if( aesni_supports( POLARSSL_AESNI_AES ) )
+    {
+        aesni_inverse_key( (unsigned char *) ctx->rk,
+                           (const unsigned char *) cty.rk, ctx->nr );
+        goto done;

```

Comment:
Heh, I'm (thankfully) not responsible for that. I just backported the changes in upstream mbedTLS (formerly PolarSSL), along with the contents of [this PR for AES-NI support on Windows](https://github.com/ARMmbed/mbedtls/pull/1355), although there appears to be a [newer replacement PR for Windows support](https://github.com/ARMmbed/mbedtls/pull/1901).

Reply:
I know this is an old PR, but as someone who has done a significant amount of C programming, I thought I should explain why this goto is there.

The reason for not having an else is to avoid code duplication without causing spaghetti code. Code duplication makes it easier to forget to free/clean up things, and spaghetti code makes code uglier and more confusing.

This is C code, not C++. In C code, forward-jumping goto statements are very commonly used for cleaning up and error handling. It is extremely common to see this in any large C code base. For example, the linux kernel has over 10k "goto" statements.